<a href="https://colab.research.google.com/github/pschorey/Valpo_IT533/blob/main/01_Processing_and_Association.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Pedagogy Lab 1 (Data Processing and Association)

>You will be working with the [STEAM Games dataset](https://raw.githubusercontent.com/shstreuber/Data-Mining/master/data/steam_games.csv) (a cleaned and modified subset of this dataset on [Kaggle](https://www.kaggle.com/datasets/tristan581/all-55000-games-on-steam-november-2022?resource=download&select=steam_games.csv). This is a PARTIAL inventory of the online PC gaming store STEAM. Parts of it were collected on the 8th of November 2022, using the STEAM store API as well as the API of another website, steamspy.com. It contains useful information, such as the products genre, tags, categories, languages and other information.
>
> The features of this dataset are
* AppID: ID of product as allocated by Steam.
* Name: Product name
* Developer: Whoever created the product.
* Publisher: Whoever published the product.
* Genre: The genre(s) that the product is in.
* Categories: The categories/features that the product has.
* Owners: An approximate number of owners, according to Steam Spy.
* Positive_Reviews: The number of positive reviews the product has.
* Negative_Reviews: The number of negativereviews the product has.
* Price_\$: The price of the game in USD.
* Initial_Price_$: The price of the game in USD at launch.
* Discount_%: What percentage sale the product was off by as of 2022/11/8.
* Peak_Concurrent_Players: Peak concurrent players as of 2022/11/8.
* Platforms: What operating systems the product is available on.
* Release Date: When the product was first released.
* Required Age: Whether the user needs to be over a certain age to legally purchase (0 = no restrictions)
* Avg_Rating_5: Average rating from 1 to 5 stars
* Playability_index_10: Average playability index on a scale from 1-10

* Source: Professor Sonja Streuber 2023S1R-IT-533-STEM:Data Mining and Applications

In [1]:
import numpy as np
import pandas as pd 
import matplotlib.pyplot as plt
import math
from math import acos, degrees
from scipy import spatial
from scipy.stats import chisquare
from scipy.stats import chi2_contingency
from sklearn.metrics.pairwise import cosine_similarity
import scipy.spatial.distance as dist #used for Manhattan dist

steam = pd.read_csv("https://raw.githubusercontent.com/pschorey/Valpo_IT533/main/steam_games_short.csv")


# Inspect the data


In [2]:
steam.head()

AppID                       Name         Developer Publisher   Genre  \
0     10             Counter-Strike             Valve     Valve  Action   
1     20      Team Fortress Classic             Valve     Valve  Action   
2     30              Day of Defeat             Valve     Valve  Action   
3     40         Deathmatch Classic             Valve     Valve  Action   
4     50  Half-Life: Opposing Force  Gearbox Software     Valve  Action   

                                          Categories  \
0  Multi-player, Valve Anti-Cheat enabled, Online...   
1  Multi-player, Valve Anti-Cheat enabled, Online...   
2             Multi-player, Valve Anti-Cheat enabled   
3  Multi-player, Valve Anti-Cheat enabled, Online...   
4  Multi-player, Single-player, Valve Anti-Cheat ...   

                     Owners  Positive_Reviews  Negative_Reviews  Price_USD  \
0  10,000,000 .. 20,000,000            201215              5199       9.99   
1    1,000,000 .. 2,000,000              5835               934       4.99   
2   5,000,000 .. 10,000,000              5251               569       4.99   
3   5,000,000 .. 10,000,000              1961               437       4.99   
4   5,000,000 .. 10,000,000             14887               756       4.99   

   Initial_Price_USD  Discount_Percent  Peak_Concurrent_Players  \
0               9.99                 0                    13990   
1               4.99                 0                      101   
2               4.99                 0                      142   
3               4.99                 0                        3   
4               4.99                 0                      106   

             Platforms Release_Date  Required_Age  Avg_Rating_5  \
0  windows, mac, linux    11/1/2000             0          3.87   
1  windows, mac, linux     4/1/1999             0          3.92   
2  windows, mac, linux     5/1/2003             0          4.90   
3  windows, mac, linux     6/1/2001             0          4.69   
4  windows, mac, linux    11/1/1999             0          2.84   

   Playability_Index_10  
0                  6.54  
1                  8.15  
2                  7.25  
3                  8.12  
4                  2.35

# Describe the data

In [3]:
steam.describe()

AppID  Positive_Reviews  Negative_Reviews     Price_USD  \
count  1.112600e+04      1.112600e+04      11126.000000  11126.000000   
mean   5.114254e+05      4.599697e+03        655.332914      8.549563   
std    5.614587e+05      6.629048e+04       9050.343291      9.223482   
min    1.000000e+01      0.000000e+00          0.000000      0.000000   
25%    2.812975e+05      2.200000e+01          9.000000      2.990000   
50%    3.658350e+05      1.105000e+02         43.000000      5.990000   
75%    4.484475e+05      6.197500e+02        173.000000      9.990000   
max    2.190950e+06      5.943345e+06     787093.000000    299.900000   

       Initial_Price_USD  Discount_Percent  Peak_Concurrent_Players  \
count       11126.000000      11126.000000             11126.000000   
mean            8.865782          3.125742               378.903200   
std             9.293126         14.635960             12008.136002   
min             0.000000          0.000000                 0.000000   
25%             2.990000          0.000000                 0.000000   
50%             6.990000          0.000000                 0.000000   
75%             9.990000          0.000000                 4.000000   
max           299.900000         90.000000            874053.000000   

       Required_Age  Avg_Rating_5  Playability_Index_10  
count  11126.000000  11126.000000          11126.000000  
mean       0.385853      3.508388              5.006447  
std        2.444761      0.864208              2.886240  
min        0.000000      2.000000              0.000000  
25%        0.000000      2.760000              2.510000  
50%        0.000000      3.520000              5.050000  
75%        0.000000      4.260000              7.480000  
max       18.000000      5.000000             10.000000

# Looking at our Steam data there aren't many categories!
We are going to create a category column based on average rating:


*   rating > 4 : "Excellent"
*   4 > rating > 3 : "Good"
*   3 > rating > 2 : "Average"
*   2 > rating : "Bad"



In [4]:
bins = [0, 2, 3, 4]
labels = ['Bad', 'Average', 'Good', 'Excellent']
d = dict(enumerate(labels, 1))
steam['Rating_Labels'] = np.vectorize(d.get)(np.digitize(steam['Avg_Rating_5'], bins))
#above code from https://stackoverflow.com/questions/49382207/how-to-map-numeric-data-into-categories-bins-in-pandas-dataframe 

#check our data to see if it worked...
steam[['Rating_Labels', 'Avg_Rating_5']]


Rating_Labels  Avg_Rating_5
0              Good          3.87
1              Good          3.92
2         Excellent          4.90
3         Excellent          4.69
4           Average          2.84
...             ...           ...
11121     Excellent          4.06
11122       Average          2.87
11123          Good          3.79
11124       Average          2.89
11125          Good          3.03

[11126 rows x 2 columns]

# How many people own each game?
The data is laid out in a way where it "guesses" the number of owners and formats it as a string.  Let's make the data more useful by creating a numerical min, max, and mid estimate of owners for each game.

In [5]:
#add numeric data for owner estimates of the games
steam['Owners_Copy'] = steam['Owners'].str.replace(',','')  #get rid of commans, https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.str.replace.html
ownersCopy = steam['Owners_Copy'].str.split(pat=' .. ', n=0, expand=True) #remove text separater, https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.str.split.html
steam['Owners_Min'] = pd.to_numeric(ownersCopy[0]) #convert string to int, https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.to_numeric.html
steam['Owners_Max'] = pd.to_numeric(ownersCopy[1])
steam['Owners_Mid'] = (steam['Owners_Min'] + steam['Owners_Max']) / 2
steam.drop(columns=['Owners_Copy'])
steam['Owners_Mid']

0        15000000.0
1         1500000.0
2         7500000.0
3         7500000.0
4         7500000.0
            ...    
11121       10000.0
11122       10000.0
11123       10000.0
11124       35000.0
11125       10000.0
Name: Owners_Mid, Length: 11126, dtype: float64

# Chi-Square

Chi Square calculates the relationship between Observed and Expected Values between two CATEGORICAL attributes.

**THINGS TO REMEMBER** about a Chi Square test:
1. Is a hypothesis test based on categorical attributes.
2. Uses as its H0 that the two variables under investigation are independent
3. Uses a chi square table
3. If the resulting p-value is > 0.05, both variables are independent
4. If the resulting p-value is < 0.05, both variables are dependent

* Source: Professor Sonja Streuber 2023S1R-IT-533-STEM:Data Mining and Applications



Using our Steam data, let's see if there is a relation between Rating_Labels and Genre.  

Our hypothesis is that game genre and reviews are dependent.

The null-hypothesis is that game genre does not affect review ratings.

Which statement is likely to be true?

In [7]:
chi2, p, dof, expected = chi2_contingency((pd.crosstab(steam.Rating_Labels, steam.Genre).values))
print (f'Chi-square Statistic: {chi2} ,p-value: {p}, Degrees of Freedom: {dof}')

Chi-square Statistic: 1411.890703428438 ,p-value: 0.48080185981389695, Degrees of Freedom: 1410


Our p-value is 0.481, which is > 0.05.  This means that there isn't an association between reviews and game genre! How dissapointing, the null-hypothesis is true.

Maybe you can find a better association?

# Now it's your turn!
Create a hypothesis using two categories from the Steam data and write it here:

**My Hypothesis is that....**

In [ ]:
#Test your hypothesis using chi2_contingency here



Summarize the results here:

**The results from my test indicate that....**

# Extra Credit 
**Evereything from this point forward is bonus, review work**

What other associations can we check?

## Euclidian Distance

In [8]:
#Euclidian Distance Example
#Is there a closer relationship between Positive_Reviews and Owners_Mid, or between Negative_Reviews and Owners_Mid?

#Using Linear Algebra
positive = np.array(steam['Positive_Reviews'])
negative = np.array(steam['Negative_Reviews'])
owners = np.array(steam['Owners_Mid'])

posDist = np.linalg.norm(owners - positive)
negDist = np.linalg.norm(owners - negative)
print("Positive review distance to owners: ", posDist, "\nNegative review distance to owners: ", negDist)

#Our analysis shows that there is a slightly closer distance between positive reviews and owners than negative reviews and owners 

Positive review distance to owners:  403400616.8212085 
Negative review distance to owners:  406539569.1973607


In [ ]:
#Your Turn: Compare using Euclidian Distance
#My Euclidian Distance Theory and example.....





# Manhattan / Cityblock Distance

In [9]:
#Manhattan distance Example
#Is there a closer relationship between Positive_Reviews and Owners_Mid, or between Negative_Reviews and Owners_Mid?
#How will this compare to the Euclidian distance?


positive = np.array(steam['Positive_Reviews'])
negative = np.array(steam['Negative_Reviews'])
owners = np.array(steam['Owners_Mid'])

print ("Manhattan Distance between Positive_Reviews and Owners_Mid:", f'{dist.cityblock(positive,owners):,}')
print ("Manhattan Distance between Negative_Reviews and Owners_Mid:", f'{dist.cityblock(negative,owners):,}')

#Similar to the Euclidian method, our analysis shows that there is a slightly closer distance between positive reviews and owners than negative reviews and owners 



Manhattan Distance between Positive_Reviews and Owners_Mid: 4,389,168,769.0
Manhattan Distance between Negative_Reviews and Owners_Mid: 4,433,053,766.0


In [ ]:
#Your Turn: Compare using Manhattan / Cityblock
#My Manhattan / Cityblock Distance Theory and example......







# Cosine Similarity

In [12]:
#Cosine Similarity Example
positive = np.array(steam['Positive_Reviews'])
rPositive = positive.reshape(1,11126)
negative = np.array(steam['Negative_Reviews'])
rNegative = negative.reshape(1,11126)
owners = np.array(steam['Owners_Mid'])
rOwners = owners.reshape(1,11126)

cosPO = cosine_similarity(rPositive,rOwners)
radPO = math.acos(cosPO)
degPO = math.degrees(radPO)

cosNO = cosine_similarity(rNegative,rOwners)
radNO = math.acos(cosNO)
degNO = math.degrees(radNO)

print("Cosine positive / owners: (cos, rad, angle)",cosPO, radPO, degPO)
print("Cosine negative / owners: (cos, rad, angle)", cosNO, radNO, degNO)

#Again we see that the negative reviews have a slightly closer association to owners than the positive reviews.

Cosine positive / owners: (cos, rad, angle) [[0.53442766]] 1.0069658733703084 57.69489465782357
Cosine negative / owners: (cos, rad, angle) [[0.5892737]] 0.9406367422652713 53.89451538673503


In [ ]:
#Your Turn: Compare using Cosine Similarity
#My Cosine Similarity Theory and example.......





# Covariance

In [14]:
#Covariance Example

posCov = np.cov([steam['Positive_Reviews']],[steam['Owners_Mid']])[0][1]
negCov = np.cov([steam['Negative_Reviews']],[steam['Owners_Mid']])[0][1]

print('Pos cov: ', posCov)
print('Neg cov: ', negCov)

Pos cov:  135232475832.38686
Neg cov:  20376638471.68086


In [17]:
#Your Turn: Compare using Covariance
#My covariance Theory and example.......



#Correlation

In [16]:
#Correlation
posCorEx = steam['Positive_Reviews'].corr(steam['Owners_Mid'])
negCorEx = steam['Negative_Reviews'].corr(steam['Owners_Mid'])

print('Positive Review correlation to owners:',posCorEx)
print('Negative Review correlation to owners:',negCorEx)

#Our results show a moderate positive correlation between reviews and owners, with Negative_Reviews having a slightly closer association

Positive Review correlation to owners: 0.5313860433255004
Negative Review correlation to owners: 0.5864726636366026


In [ ]:
#Your Turn: Compare using Correlation
#My correlation Theory and example.......

